### PR TITLE
Fixed checkbox value not changing

### PR DESCRIPTION
**Pull request by mikededo**
So, what was happening is that you were trying to change the `UnmodifiableList` from the `Filter`. The `Filter` was declared as a `final` variable which means **you can't** change it's properties (for more information, visit the [api](https://dart.dev/guides/language/language-tour#final-and-const)).

What I've done is, I created two methods that returned a new `Filter` object instance which is the new updated filter. However, I do not belive these are good practices. I would recommend you created two new methods that updated the `Filter` object from the inside of the class (one of the principles of OOP). This would allow you to create more secure code. 

To sum up, I made a bit of code refactoring, I hope you don't mind it. You can always revert the changes 😉. I also added a `BlocObserver` which is a nice tool to debug your app since it tells you when an event is thrown and the state changes, as well as the errors 💯.

### CODE REVIEW

**Comment by rravithejareddy** [comment]
Thanks now you got my point, this is what I was looking for, in your answer you are saying that creating new object is bad that I can overcome with equtable, that's what I was trying to say you, and I wanted reply from felangal on this code whether I am doing right or wrong, but your code is good, except creating new object.

**Comment by mikededo** [comment]
But it is because you are not using the bloc properly. I do not really see the point of the `Filter` class. Why are you keeping the items in lists? Why are you passing a `Filter` object in the `CheckBoxList`? You should use `BlocProvider.of...` to get the state filter. If you check the examples in the bloclibrady.dev (which I believe you already have) there's a _FilteredTodosBloc_ in the _Todos_ example. I would strongly recommend you to give it a look. Also, I would recommend you to pay more attention to the debug output of your IDE since the solution of your problem was explained in it.

If you want your code to be reviewed, you shoudl write more explanatory issues since it was hard to understand. 😃 

If there's anything else, hit me up.

**Comment by mikededo** [comment]
By the way, does it work as expected?
